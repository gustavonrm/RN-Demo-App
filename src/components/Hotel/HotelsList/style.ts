import { StyleSheet } from 'react-native';

const style = StyleSheet.create({
  container: {
    marginHorizontal: 20,
  },
});

export default style;
