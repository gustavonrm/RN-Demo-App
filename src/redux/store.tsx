import { configureStore, combineReducers } from '@reduxjs/toolkit';
import hotelsSlice from './slices/hotels.slice';
import { hotelsApi } from './apis/hotels.api';
import filtersSlice from './slices/filters.slice';

const rootReducer = combineReducers({
  hotels: hotelsSlice,
  filters: filtersSlice,
  [hotelsApi.reducerPath]: hotelsApi.reducer,
});

export const setupStore = (preloadedState?: Partial<RootState>) =>
  configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(hotelsApi.middleware),
    preloadedState,
  });

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
