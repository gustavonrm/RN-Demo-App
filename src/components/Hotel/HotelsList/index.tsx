import React, { useMemo } from 'react';
import { FlatList } from 'react-native';
import { Hotel } from '../types/hotel';
import HotelCard from '../HotelCard';
import withLoader from '../../HOC/withLoader';
import { useGetHotelsQuery } from '../../../redux/apis/hotels.api';
import style from './style';
import withError from '../../HOC/withError';
import { selectHotels } from '../../../redux/slices/hotels.slice';
import { useSelector } from 'react-redux';
import { selectFilters } from '../../../redux/slices/filters.slice';
import ErrorView from '../../common/ErrorView';

type Filters = {
  stars?: number;
  price?: string;
};

const HotelsList = () => {
  const hotels = useSelector(selectHotels);
  const filters = useSelector(selectFilters) as Filters;
  // const [filtering, setFiltering] = useState(false); // Not added because the filtering is very fast and the modal also hides this
  const filteredHotels = useMemo(() => {
    const filterKeys = ['stars', 'price'];
    const hasFilters = filterKeys.some((key) => key in filters);

    let filtered = hotels.filter((hotel) => {
      if (!hasFilters) {
        return true;
      }

      const matchesStars = filters.stars !== undefined ? hotel.stars === filters.stars : true;

      return matchesStars;
    });

    if (filters.price) {
      filtered = filtered.sort((a: Hotel, b: Hotel) => {
        if (filters.price === 'desc') {
          return b.price - a.price;
        } else if (filters.price === 'asc') {
          return a.price - b.price;
        }
        return 0;
      });
    }

    return filtered;
  }, [hotels, filters]);

  const renderItem = ({ item }: { item: Hotel }) => {
    return <HotelCard hotel={item} />;
  };

  return (
    <FlatList
      contentContainerStyle={{ flexGrow: 1 }}
      style={style.container}
      showsVerticalScrollIndicator={false}
      data={filteredHotels}
      renderItem={renderItem}
      keyExtractor={(item) => item.id}
      ListEmptyComponent={<ErrorView message={'No hotels found'} />}
    />
  );
};

export default withLoader(withError(HotelsList, selectHotels), useGetHotelsQuery);
