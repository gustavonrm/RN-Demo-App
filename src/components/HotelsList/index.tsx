import React from 'react';
import { FlatList } from 'react-native';
import { Hotel } from '../types/hotel';
import HotelCard from '../HotelCard';
import withLoader from '../HOC/withLoader';
import { useGetHotelsQuery } from '../../redux/apis/hotels.api';
import style from './style';

type HotelListProps = {
  data: Hotel[];
};

const HotelsList = ({ data }: HotelListProps) => {
  const renderItem = ({ item }: { item: Hotel }) => <HotelCard hotel={item} />;

  return (
    <FlatList
      style={style.container}
      showsVerticalScrollIndicator={false}
      data={data}
      renderItem={renderItem}
      keyExtractor={(item) => item.id}
    />
  );
};

export default withLoader(HotelsList, useGetHotelsQuery);
