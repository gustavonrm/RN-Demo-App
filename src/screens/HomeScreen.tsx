import React from 'react';
import HotelsList from '../components/HotelsList';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import { View } from 'react-native';
import Filter from '../components/common/Filter';

const HomeScreen = () => {
  const insets = useSafeAreaInsets();
  const styleSafeArea = { flex: 1, paddingLeft: insets.left, paddingRight: insets.right };

  return (
    <View style={styleSafeArea}>
      <Filter>
        <Filter.Toggle />
        <Filter.Menu />
      </Filter>
      <HotelsList />
    </View>
  );
};

export default HomeScreen;
