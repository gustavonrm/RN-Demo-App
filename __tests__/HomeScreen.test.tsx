import React from 'react';
import 'react-native';
import {
  cleanup,
  screen,
  waitForElementToBeRemoved,
  userEvent,
} from '@testing-library/react-native';
import HotelsList from '../src/components/Hotel/HotelsList';
import { renderWithProviders } from './utils/test-utils';
import { http, HttpResponse, delay } from 'msw';
import { setupServer } from 'msw/native';
import { hotels } from './mocks/hotels';
import HomeScreen from '../src/screens/HomeScreen';
import { setupStore } from '../src/redux/store';
import { SafeAreaProvider, useSafeAreaInsets } from 'react-native-safe-area-context';

jest.useFakeTimers();

export const handlers = [
  http.get('https://run.mocky.io/v3/b4593f68-9e29-47a8-bb6d-8137a665ee15', async () => {
    await delay(150);
    return HttpResponse.json(hotels);
  }),
];

export const server = setupServer(...handlers);

beforeAll(() => {
  server.listen();
});

afterEach(() => {
  server.resetHandlers();
  jest.restoreAllMocks();
  cleanup();
});

afterAll(() => {
  server.close();
});

describe('HotelList Element', () => {
  it('renders hotel list correctly', async () => {
    const store = setupStore();

    renderWithProviders(<HomeScreen />, { store });
    expect(screen.getByTestId('loaderTestId')).toBeTruthy();
    await waitForElementToBeRemoved(() => screen.getByTestId('loaderTestId'));

    expect(await screen.getAllByTestId('hotelCardTestId').length).toBe(10);
  });

  it('renders filters by star correctly', async () => {
    const user = userEvent.setup();
    const store = setupStore();

    renderWithProviders(<HomeScreen />, { store });
    expect(screen.getByTestId('loaderTestId')).toBeTruthy();
    await waitForElementToBeRemoved(() => screen.getByTestId('loaderTestId'));

    await user.press(screen.getByTestId('filterButtonTestId'));
    await user.press(screen.getByTestId('5StarsFilterTestId'));
    await user.press(screen.getByTestId('applyFilterButtonTestId'));

    expect(await screen.getAllByTestId('hotelCardTestId').length).toBe(3);
  });

  it('renders filters sort by price decrease correctly', async () => {
    const user = userEvent.setup();
    const store = setupStore();

    renderWithProviders(<HomeScreen />, { store });
    expect(screen.getByTestId('loaderTestId')).toBeTruthy();
    await waitForElementToBeRemoved(() => screen.getByTestId('loaderTestId'));

    await user.press(screen.getByTestId('filterButtonTestId'));
    await user.press(screen.getByTestId('priceDescFilterTestId'));
    await user.press(screen.getByTestId('applyFilterButtonTestId'));

    const hotelCards = screen.getAllByTestId('priceTestId');

    for (let i = 1; i < hotelCards.length; i++) {
      let price1 = parseFloat(hotelCards[i - 1].children[0].toString().replace('€', ''));
      let price2 = parseFloat(hotelCards[i].children[0].toString().replace('€', ''));
      expect(price1).toBeGreaterThanOrEqual(price2);
    }
  });

  it('renders filters filter by average rating correctly', async () => {
    const user = userEvent.setup();
    const store = setupStore();

    renderWithProviders(<HomeScreen />, { store });
    expect(screen.getByTestId('loaderTestId')).toBeTruthy();
    await waitForElementToBeRemoved(() => screen.getByTestId('loaderTestId'));

    await user.press(screen.getByTestId('filterButtonTestId'));
    await user.press(screen.getByTestId('ratingAvgFilterTestId'));
    await user.press(screen.getByTestId('applyFilterButtonTestId'));

    expect(await screen.getAllByTestId('hotelCardTestId').length).toBe(2);
  });

  it('renders filters sort by price decrease and 4 stars correctly', async () => {
    const user = userEvent.setup();
    const store = setupStore();

    renderWithProviders(<HomeScreen />, { store });
    expect(screen.getByTestId('loaderTestId')).toBeTruthy();
    await waitForElementToBeRemoved(() => screen.getByTestId('loaderTestId'));

    await user.press(screen.getByTestId('filterButtonTestId'));
    await user.press(screen.getByTestId('4StarsFilterTestId'));
    await user.press(screen.getByTestId('priceDescFilterTestId'));
    await user.press(screen.getByTestId('applyFilterButtonTestId'));

    expect(await screen.getAllByTestId('hotelCardTestId').length).toBe(6);
    const hotelCards = screen.getAllByTestId('priceTestId');

    for (let i = 1; i < hotelCards.length; i++) {
      let price1 = parseFloat(hotelCards[i - 1].children[0].toString().replace('€', ''));
      let price2 = parseFloat(hotelCards[i].children[0].toString().replace('€', ''));
      expect(price1).toBeGreaterThanOrEqual(price2);
    }
  });

  it('snapshot test', () => {
    const store = setupStore();
    expect(renderWithProviders(<HomeScreen />, { store })).toMatchSnapshot();
  });
});
