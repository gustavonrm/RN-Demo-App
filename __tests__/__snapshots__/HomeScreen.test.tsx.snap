// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotelList Element snapshot test 1`] = `
<View
  style={
    {
      "flex": 1,
      "paddingLeft": 0,
      "paddingRight": 0,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "bottom": 50,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "zIndex": 1,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#E12C70",
          "borderRadius": 25,
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 5,
          "zIndex": 2,
        }
      }
      testID="filterButtonTestId"
    >
      <View />
      <Text
        style={
          {
            "color": "#FFF",
            "fontSize": 16,
            "fontWeight": "bold",
            "marginLeft": 5,
          }
        }
      >
        Filter
      </Text>
    </View>
  </View>
  <ActivityIndicator
    color="#E12C70"
    size="large"
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
    testID="loaderTestId"
  />
</View>
`;
